src/data/list/cycle.lean

```diff
@@ -699,4 +713,103 @@ underlying element. This representation also supports cycles that can contain du
 instance [has_repr α] : has_repr (cycle α) :=
 ⟨λ s, "c[" ++ string.intercalate ", " ((s.map repr).lists.sort (≤)).head ++ "]"⟩
 
+/-- `chain R s` means that `R` holds between adjacent elements of `s`.
+
+     chain R [a, b, c] ↔ R a b ∧ R b c ∧ R c a -/
+def chain (r : α → α → Prop) : cycle α → Prop :=
+@quotient.lift _ _ (is_rotated.setoid α) (λ l, match l with
```

Comment:
The primed versions save you some underscores:
```suggestion
def chain (r : α → α → Prop) (c : cycle α) : Prop :=
quotient.lift_on' c (λ l, match l with
```

Reply:
I assume that comment was meant for the line below, @pechersky?

Reply:
Correct

Reply:
Actually, more briefly, I'd argue that in a singleton, the unique element is adjacent to itself.

Reply:
We still support natively non-nodup cycles of `[x, x]` also. Might or might not be relevant for what you'll be using this for.

If I had a graph of a single vertex, then by convention, it contains a cycle:
https://mathworld.wolfram.com/SingletonGraph.html

Right now, if your `r = G.adj`, then we cannot encode this using `cycle`, using your current `chain` definition.

Reply:
This to me seems like a case of two different conventions which are both useful in their specific contexts. The cyclic graph definition could still use `pairwise`, either by taking the reflexive closure (to allow walks that don't move you anywhere), or by simply hardcoding the singleton case.

If it really comes down to it, we could define a `chain'` predicate that includes this case, but I'm almost certain that what I really need for my use case is what I currently have.

Reply:
I think I'm convinced that @vihdzp has the right definition here, but it might be worth asking on zulip.

I don't think appealing to `simple_graph.adj` helps much, because it's not at all clear that that's defined in a sensible way regarding single-vertex graphs.